Feature/signals

ybastide opened the pull request:
Based on child workflows, which really must be merged :wink:

jbbarth commented on simpleflow/swf/executor.py:
```diff
@@ -524,18 +531,13 @@ def schedule_task(self, a_task, task_list=None):
         # TODO: if we had exactly MAX_DECISIONS - 1 to take, this will wake up
         # the workflow for no reason. Evaluate if we can do better.
-        if len(self._decisions) == constants.MAX_DECISIONS - 1:
+        if len(self._decisions) >= constants.MAX_DECISIONS - 1:
```
Note sure the `>=` can happen :p

ybastide replied:
Not currently as `decisions` is a 1-item list :wink:
But `SignalTask.schedule` could have scheduled the signal on the children and parent workflow too, who knows?

jbbarth replied:
Hm OK. Because if we send > 100 decisions in a RespondDecisionTaskCompleted call, I think SWF will yell at us. Do you think it could happen?

ybastide replied:
D'oh: you're right, of course.

jbbarth commented on simpleflow/history.py:
```diff
@@ -36,12 +39,21 @@ def activities(self):
         return self._child_workflows
 
     @property
+    @property
```
Double `@property` here?

ybastide replied:
Oops

jbbarth commented:
Looks nice! Just to be sure I understand: we keep the signals in a (ordered) dict so it means in case that 2 signals are received with the same name but different values, only the last one will be processed ? I'm fine with that for our current use cases, but I wonder if there's some unicity constraint dictated by SWF limits or it's just a choice of our own ?

I think we could merge that soon. If you want something we can improve (later?) : I'd love to have better examples, the ones here are not easy to grok imho. Maybe we can imagine some real world task, let's think about it :)

ybastide commented:
"2 signals are received with the same name but different values": well, I didn't even think about this... I imagined signal names to be unique, but this isn't needed.

jbbarth commented:
That's a tradeoff/assumption we can make if it's simpler. We just have to be careful the latest signal is the one processed (in our case: pausing/resuming a crawl work with that, but we use different signal names and no args so no problem imho). And we'll change if somebody complains, yay?

ybastide commented:
* added a skeleton doc: simpleflow/docs/signals.rst
* added the ability to specify a workflow to signal
* fixed? broadcasting

Note: example 2 doesn't seem to work in sandalone mode

jbbarth commented:
👍  on merging this, but just a few questions before:
- do we want to keep all the comments you left in the code? (various `logger.foo()` commented out, and some dead code it seems too)
- from a pure esthetic perspective, I find it weird to use `self.submit()` for sending and waiting for a signal. Is there a specific reason for doing it that way? I guess we could have just `self.signal("...")` or `self.wait_signal("...")` directly?
- not sure I can understand all the conditions for signal propagation, but we'll see how it goes in action :-)

Good job anyway, and THANKS YOU SO MUCH for writing some docs. We should do this more often. I started exploring how to build a docs site btw, I should have something to show in the next few weeks.

ybastide commented:
* dead code: you're right, I'll remove all this cruft 🙂 
* on using `submit`: mm... The thing is, we want to put `signal`/`wait_signal` inside `Group`s; so they should  return something "submittable" (or Group could handle them specifically); but more importantly they must be asynchronous: `Chain(a, signal1, b)` mustn't signal before `a`.
    Can this be done in another way? Should we define convenience methods to hide `submit`, or inversely rename the current `signal`/`wait_signal`?
* signal propagation (and I'll rename `process` to `propagate`, much better!): We want to send the signal to our parent and started children, except those we already signaled and the original sender.
    This original sender is known either via `external_*`, for a signal sent with SignalExternalWorkflow, or our own `__*_id` for signals sent with SignalWorkflowExecution (SignalWorkflowExecution doesn't set the  `external_*` variables and I wasn't able to reliably use SignalExternalWorkflow... Jobs ending before the signal could be sent or received, it's hazy now 😞)
    Oh, and we may want another `propagate` option: children only (parent(s) only should be doable already)

jbbarth commented:
Hm, didn't think about chains, this totally makes sense now. Thanks for the additional infos!
